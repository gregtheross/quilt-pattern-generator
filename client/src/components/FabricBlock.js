import React from "react";

class FabricBlock extends React.Component {
  render() {
    const style = {
      clipPath: "url(#" + this.props.id + ")",
      position: "absolute",
      top: this.props.top,
      left: this.props.left
    };

    return (
      <div>
        <img
          style={style}
          src={this.props.backgroundImage}
          width={this.props.width}
          height={this.props.height}
        />
        <svg>
          <defs>
            <clipPath id={this.props.id}>
              <polygon points={this.props.points} />
            </clipPath>
          </defs>
        </svg>
      </div>
    );
  }
}

export default FabricBlock;
