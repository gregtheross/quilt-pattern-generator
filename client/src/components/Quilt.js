import React from "react";
import SvgTriangle from "./SvgTriangle.js";

class Quilt extends React.Component {
  // todo: pass a list of fabric entities (could simply be an id and url) into the props
  // todo: shuffle the fabrics

  getFabricList() {
    // Generate initial array of fabricIds and then shuffle them.
    // This is purposeful because we want an even number of fabric swatches to be used and only their locations randomized.
    // If we randomize the selection there is a high risk of having more fabric blocks than others
    let indexes = [];
    for (let i = 0; i < this.props.colCount * this.props.rowCount; i++) {
      indexes.push((i % this.props.fabricList.length));
    }

    var currentIndex = indexes.length,
      temporaryValue,
      randomIndex;

    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;

      // And swap it with the current element.
      temporaryValue = indexes[currentIndex];
      indexes[currentIndex] = indexes[randomIndex];
      indexes[randomIndex] = temporaryValue;
    }

    return indexes;
  }

  createSvgQuilt() {
    // todo: eventually this should get handle different calculations based on the selected shape

    let shuffledFabrics = this.getFabricList();

    let triangleWidth = this.props.shapeWidth;
    let triangleHeight = this.props.shapeHeight > 0 ? this.props.shapeHeight : triangleWidth * Math.sqrt(3) / 2;

    let trianglePointsDown = "0 0, " + triangleWidth / 2 + " " + triangleHeight + ", " + triangleWidth + " 0";
    let trianglePointsUp = "" + triangleWidth / 2 + " 0, " + triangleWidth + " " + triangleHeight + ", 0 " + triangleHeight;

    let shapeIndex = 0;

    let svgShapes = [];
    let isUp = true;

    for (let rowIndex = 0; rowIndex < this.props.rowCount; rowIndex++) {
      isUp = rowIndex % 2 === 0;
      for (let colIndex = 0; colIndex < this.props.colCount; colIndex++) {
        svgShapes.push(
          <SvgTriangle
            id={shapeIndex}
            key={shapeIndex}
            points={isUp ? trianglePointsUp : trianglePointsDown}
            width={triangleWidth}
            height={triangleHeight}
            top={rowIndex * triangleHeight}
            left={(colIndex * triangleWidth) / 2}
            backgroundImage={this.props.fabricList[shuffledFabrics[shapeIndex]]}
          />
        );

        isUp = !isUp;
        shapeIndex++;
      }
    }
    return svgShapes;
  }

  render() {
    return (
      <div>
        <div className="svg-quilt">{this.createSvgQuilt()}</div>
      </div>
    );
  }
}

export default Quilt;
