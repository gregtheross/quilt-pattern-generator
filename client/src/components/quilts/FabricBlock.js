import React from "react";
import PropTypes from "prop-types";

class FabricBlock extends React.Component {
  handleClick = e => {
    this.props.onFabricBlockClick(this.props.id);
  };

  render() {
    const clipPathId = `${this.props.clipPathPrefix}-${this.props.id}`;

    const style = {
      clipPath: "url(#" + clipPathId + ")",
      position: "absolute",
      top: this.props.top,
      left: this.props.left,
      opacity: this.props.selected ? ".25" : "1"
    };

    return (
      <div>
        <img
          alt={`tile ${this.props.id}`}
          style={style}
          src={this.props.backgroundImage}
          width={this.props.width}
          height={this.props.height}
          onClick={this.handleClick}
        />
        <svg>
          <defs>
            <clipPath id={clipPathId}>
              <polygon points={this.props.points} />
            </clipPath>
          </defs>
        </svg>
      </div>
    );
  }
}

FabricBlock.propTypes = {
  backgroundImage: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  id: PropTypes.number.isRequired,
  points: PropTypes.string.isRequired,
  onFabricBlockClick: PropTypes.func.isRequired,
  selected: PropTypes.bool.isRequired,
  clipPathPrefix: PropTypes.string.isRequired
};

export default FabricBlock;
