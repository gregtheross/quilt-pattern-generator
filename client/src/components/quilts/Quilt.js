import React from "react";
import FabricBlock from "./FabricBlock.js";
import PropTypes from "prop-types";

class Quilt extends React.Component {
  createQuilt() {
    if (
      this.props.fabricList &&
      this.props.quiltBlocks &&
      this.props.fabricList.length > 0 &&
      this.props.quiltBlocks.length > 0
    ) {
      switch (this.props.shapeType) {
        case 1:
          return this.createTriangleQuilt(true);
        case 2:
          return this.createTriangleQuilt(false);
        case 3:
          return this.createSquareQuilt();
        case 4:
          return this.createHexagonQuilt();
        default:
          return <div>Invalid Shape</div>;
      }
    }
  }

  createTriangleQuilt(isEquilateral) {
    let triangleWidth = this.props.shapeWidth;
    // if no height entered, use equilateral triangle calculation
    let triangleHeight = isEquilateral
      ? (triangleWidth * Math.sqrt(3)) / 2
      : this.props.shapeHeight;

    // since the triangles alternate, the coordinates will always be the same for when they point up or down
    let trianglePointsDown = `0 0, ${triangleWidth /
      2} ${triangleHeight}, ${triangleWidth} 0`;
    let trianglePointsUp = `${triangleWidth /
      2} 0, ${triangleWidth} ${triangleHeight}, 0 ${triangleHeight}`;

    let shapeIndex = 0;

    let svgShapes = [];
    let isUp = true;

    for (let rowIndex = 0; rowIndex < this.props.rowCount; rowIndex++) {
      isUp = rowIndex % 2 === 0;
      for (let colIndex = 0; colIndex < this.props.colCount; colIndex++) {
        let currentFabric = this.props.fabricList.find(
          // eslint-disable-next-line
          fabric => fabric.id === this.props.quiltBlocks[shapeIndex]
        );

        svgShapes.push(
          <FabricBlock
            id={shapeIndex}
            key={shapeIndex}
            points={isUp ? trianglePointsUp : trianglePointsDown}
            width={triangleWidth}
            height={triangleHeight}
            top={rowIndex * triangleHeight}
            left={(colIndex * triangleWidth) / 2}
            backgroundImage={currentFabric && currentFabric.url}
            selected={shapeIndex === this.props.selectedBlockIndex}
            onFabricBlockClick={this.props.onFabricBlockClick}
            clipPathPrefix="quilt"
          />
        );

        isUp = !isUp;
        shapeIndex++;
      }
    }
    return svgShapes;
  }

  createSquareQuilt() {
    let squareWidth = this.props.shapeWidth;
    let squareHeight = this.props.shapeWidth;

    let squarePoints = `0 0, ${squareWidth} 0, ${squareWidth} ${squareHeight}, 0 ${squareHeight}`;

    let shapeIndex = 0;

    let svgShapes = [];

    for (let rowIndex = 0; rowIndex < this.props.rowCount; rowIndex++) {
      for (let colIndex = 0; colIndex < this.props.colCount; colIndex++) {
        let currentFabric = this.props.fabricList.find(
          // eslint-disable-next-line
          fabric => fabric.id === this.props.quiltBlocks[shapeIndex]
        );

        svgShapes.push(
          <FabricBlock
            id={shapeIndex}
            key={shapeIndex}
            points={squarePoints}
            width={squareWidth}
            height={squareHeight}
            top={rowIndex * squareHeight}
            left={colIndex * squareWidth}
            backgroundImage={currentFabric && currentFabric.url}
            selected={shapeIndex === this.props.selectedBlockIndex}
            onFabricBlockClick={this.props.onFabricBlockClick}
            clipPathPrefix="quilt"
          />
        );

        shapeIndex++;
      }
    }
    return svgShapes;
  }

  createHexagonQuilt() {
    // calculate some re-usable lengths broken up from the hexagon
    let centerlineWidth = this.props.shapeWidth;
    let heightOffCenter = (centerlineWidth / 4) * Math.sqrt(3);
    let internalTriangleShortSide = centerlineWidth / 4;

    let hexagonWidth = centerlineWidth;
    let hexagonHeight = heightOffCenter * 2;

    let hexagonPoints =
      `0 ${heightOffCenter},` +
      ` ${internalTriangleShortSide} 0,` +
      ` ${internalTriangleShortSide * 3} 0,` +
      ` ${centerlineWidth} ${heightOffCenter}` +
      ` ${internalTriangleShortSide * 3} ${heightOffCenter * 2},` +
      ` ${internalTriangleShortSide} ${heightOffCenter * 2}`;

    let shapeIndex = 0;

    let svgShapes = [];
    let offsetCol;

    for (let rowIndex = 0; rowIndex < this.props.rowCount; rowIndex++) {
      offsetCol = false;

      for (let colIndex = 0; colIndex < this.props.colCount; colIndex++) {
        let currentFabric = this.props.fabricList.find(
          // eslint-disable-next-line
          fabric => fabric.id === this.props.quiltBlocks[shapeIndex]
        );

        svgShapes.push(
          <FabricBlock
            id={shapeIndex}
            key={shapeIndex}
            points={hexagonPoints}
            width={hexagonWidth}
            height={hexagonHeight}
            top={rowIndex * hexagonHeight + (offsetCol ? hexagonHeight / 2 : 0)}
            left={colIndex * internalTriangleShortSide * 3}
            backgroundImage={currentFabric && currentFabric.url}
            selected={shapeIndex === this.props.selectedBlockIndex}
            onFabricBlockClick={this.props.onFabricBlockClick}
            clipPathPrefix="quilt"
          />
        );

        offsetCol = !offsetCol;
        shapeIndex++;
      }
    }
    return svgShapes;
  }

  render() {
    return (
      <div>
        <div className="svg-quilt">{this.createQuilt()}</div>
      </div>
    );
  }
}

Quilt.propTypes = {
  shapeHeight: PropTypes.number.isRequired,
  shapeWidth: PropTypes.number.isRequired,
  rowCount: PropTypes.number.isRequired,
  colCount: PropTypes.number.isRequired,
  fabricList: PropTypes.array.isRequired,
  selectedBlockIndex: PropTypes.number,
  onFabricBlockClick: PropTypes.func.isRequired,
  quiltBlocks: PropTypes.array.isRequired,
  shapeType: PropTypes.number.isRequired
};

export default Quilt;
