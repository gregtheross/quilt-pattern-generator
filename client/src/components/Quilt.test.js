import React from "react";
import { shallow, mount } from "enzyme";
import Quilt from "./Quilt";
import FabricBlock from "./FabricBlock";

it("should pass", () => {
  expect(true).toEqual(true);
});

function renderQuilt(args) {
  const defaultProps = {
    rowCount: 1,
    colCount: 1,
    shapeType: 1,
    shapeWidth: 100,
    shapeHeight: 100,
    fabricList: [{ id: 1, url: "" }],
    quiltBlocks: [1],
    onFabricBlockClick: jest.fn()
  };

  const props = { ...defaultProps, ...args };

  return mount(<Quilt {...props} />);
}

it("100x100 equilateral triangle", () => {
  const wrapper = renderQuilt();
  //   console.log(wrapper.debug());
  const expextedPolygon = (
    <polygon points="50 0, 100 86.60254037844386, 0 86.60254037844386" />
  );

  expect(wrapper.contains(expextedPolygon)).toEqual(true);
});

"square",
  it("80x100 isosceles triangle", () => {
    const wrapper = renderQuilt({
      shapeType: 2,
      shapeHeight: 80
    });
    // console.log(wrapper.debug());
    const expextedPolygon = <polygon points="50 0, 100 80, 0 80" />;

    expect(wrapper.contains(expextedPolygon)).toEqual(true);
  });

it("100x100 square", () => {
  const wrapper = renderQuilt({ shapeType: 3 });
  //   console.log(wrapper.debug());
  const expextedPolygon = <polygon points="0 0, 100 0, 100 100, 0 100" />;

  expect(wrapper.contains(expextedPolygon)).toEqual(true);
});

it("100x100 hexagon", () => {
  const wrapper = renderQuilt({ shapeType: 4 });
  //   console.log(wrapper.debug());
  const expextedPolygon = (
    <polygon points="0 43.30127018922193, 25 0, 75 0, 100 43.30127018922193 75 86.60254037844386, 25 86.60254037844386" />
  );

  expect(wrapper.contains(expextedPolygon)).toEqual(true);
});
