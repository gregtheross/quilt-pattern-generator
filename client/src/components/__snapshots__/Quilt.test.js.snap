// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`80x100 isosceles triangle 1`] = `
<div>
  <div
    className="svg-quilt"
  >
    <div>
      <img
        alt="tile 0"
        height={80}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#0)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 0,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={0}
          >
            <polygon
              points="50 0, 100 80, 0 80"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 1"
        height={80}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#1)",
            "left": 50,
            "opacity": "1",
            "position": "absolute",
            "top": 0,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={1}
          >
            <polygon
              points="0 0, 50 80, 100 0"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 2"
        height={80}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#2)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 80,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={2}
          >
            <polygon
              points="0 0, 50 80, 100 0"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 3"
        height={80}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#3)",
            "left": 50,
            "opacity": "1",
            "position": "absolute",
            "top": 80,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={3}
          >
            <polygon
              points="50 0, 100 80, 0 80"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
  </div>
</div>
`;

exports[`100x100 equilateral triangle 1`] = `
<div>
  <div
    className="svg-quilt"
  >
    <div>
      <img
        alt="tile 0"
        height={86.60254037844386}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#0)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 0,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={0}
          >
            <polygon
              points="50 0, 100 86.60254037844386, 0 86.60254037844386"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 1"
        height={86.60254037844386}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#1)",
            "left": 50,
            "opacity": "1",
            "position": "absolute",
            "top": 0,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={1}
          >
            <polygon
              points="0 0, 50 86.60254037844386, 100 0"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 2"
        height={86.60254037844386}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#2)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 86.60254037844386,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={2}
          >
            <polygon
              points="0 0, 50 86.60254037844386, 100 0"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 3"
        height={86.60254037844386}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#3)",
            "left": 50,
            "opacity": "1",
            "position": "absolute",
            "top": 86.60254037844386,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={3}
          >
            <polygon
              points="50 0, 100 86.60254037844386, 0 86.60254037844386"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
  </div>
</div>
`;

exports[`100x100 hexagon 1`] = `
<div>
  <div
    className="svg-quilt"
  >
    <div>
      <img
        alt="tile 0"
        height={86.60254037844386}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#0)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 0,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={0}
          >
            <polygon
              points="0 43.30127018922193, 25 0, 75 0, 100 43.30127018922193 75 86.60254037844386, 25 86.60254037844386"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 1"
        height={86.60254037844386}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#1)",
            "left": 75,
            "opacity": "1",
            "position": "absolute",
            "top": 43.30127018922193,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={1}
          >
            <polygon
              points="0 43.30127018922193, 25 0, 75 0, 100 43.30127018922193 75 86.60254037844386, 25 86.60254037844386"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 2"
        height={86.60254037844386}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#2)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 86.60254037844386,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={2}
          >
            <polygon
              points="0 43.30127018922193, 25 0, 75 0, 100 43.30127018922193 75 86.60254037844386, 25 86.60254037844386"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 3"
        height={86.60254037844386}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#3)",
            "left": 75,
            "opacity": "1",
            "position": "absolute",
            "top": 129.9038105676658,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={3}
          >
            <polygon
              points="0 43.30127018922193, 25 0, 75 0, 100 43.30127018922193 75 86.60254037844386, 25 86.60254037844386"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
  </div>
</div>
`;

exports[`100x100 square 1`] = `
<div>
  <div
    className="svg-quilt"
  >
    <div>
      <img
        alt="tile 0"
        height={100}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#0)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 0,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={0}
          >
            <polygon
              points="0 0, 100 0, 100 100, 0 100"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 1"
        height={100}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#1)",
            "left": 100,
            "opacity": "1",
            "position": "absolute",
            "top": 0,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={1}
          >
            <polygon
              points="0 0, 100 0, 100 100, 0 100"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 2"
        height={100}
        onClick={[Function]}
        src="2.jpg"
        style={
          Object {
            "clipPath": "url(#2)",
            "left": 0,
            "opacity": "1",
            "position": "absolute",
            "top": 100,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={2}
          >
            <polygon
              points="0 0, 100 0, 100 100, 0 100"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
    <div>
      <img
        alt="tile 3"
        height={100}
        onClick={[Function]}
        src="1.jpg"
        style={
          Object {
            "clipPath": "url(#3)",
            "left": 100,
            "opacity": "1",
            "position": "absolute",
            "top": 100,
          }
        }
        width={100}
      />
      <svg>
        <defs>
          <clipPath
            id={3}
          >
            <polygon
              points="0 0, 100 0, 100 100, 0 100"
            />
          </clipPath>
        </defs>
      </svg>
    </div>
  </div>
</div>
`;
